sampled version of sparse_softmax_cross_entropy_with_logits

TTrapper opened the pull request:
Following from the already implemented `sampled_softmax_loss`, this is a very similar implementation that performs sampling for `sparse_softmax_cross_entropy_with_logits`.

TTrapper committed:
sampled version of sparse_softmax_cross_entropy_with_logits

tensorflow-jenkins commented:
Can one of the admins verify this patch?

frankchn commented:
Jenkins, test this please.

drpngx commented:
@ebrevdo WDYT? Should we do this in contrib first?

TTrapper commented:
Hi folks, thank you for looking at this request. In regards to the test failures:

- **Linux CPU Tests:** `api_compatibility_test`
I take it this is to be expected given that the API would change. If these changes are accepted, how do I go about updating the goldens?

- **Linux CPU Tests (Python 3):** ` do_virtualenv_pip_test`
Not sure about this one, but I can investigate tomorrow.

ebrevdo commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -1258,3 +1258,101 @@ def sampled_softmax_loss(weights,
+    logits = tf.matmul(inputs, tf.transpose(weights))
+    logits = tf.nn.bias_add(logits, biases)
+    labels_one_hot = tf.one_hot(labels, n_classes)
+    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
```
The labels to this function are not one hot.

TTrapper replied:
Thank you. Corrected.

ebrevdo commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -1258,3 +1258,101 @@ def sampled_softmax_loss(weights,
+  ([pdf](http://arxiv.org/pdf/1412.2007.pdf)) for the math.
+
+  Args:
+    weights: A `Tensor` of shape `[num_classes, dim]`, or a list of `Tensor`
```
Tensor or variable?

TTrapper replied:
Certainly a Variable, but this is the terminology used in the documentation of similar functions (eg `sampled_softmax_loss`). I take it that the term Tensor is in inclusive of variable Tensors? Let me know if you would still like it changed.

ebrevdo commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -1258,3 +1258,101 @@ def sampled_softmax_loss(weights,
+      partition_strategy=partition_strategy,
+      name=name)
+
+  # labels returned by _compute_sampled_logits are one_hot. Convert to indices.
```
Seems wasteful, perhaps that function can be modified to optionally return labels as integers.

TTrapper replied:
OK, I added an option to do just that.

ebrevdo reviewed (changes requested):
This is a reasonable candidate for tf.contrib.nn, but not core until it gets some use.

TTrapper committed:
Adressed reviewer comments: moved to contrib, fixed erroneous doc, modified _compute_sampled_logits to optionally return target indices

TTrapper committed:
checkstyle fix

TTrapper reviewed (commented):
Requested changes have been implemented/addressed.

TTrapper commented:
@frankchn I am new to this. Is there something I should do to have the `stat: awaiting response` label removed? Thanks!

frankchn commented:
Jenkins, test this please.

frankchn commented:
@ebrevdo PTAL again. Thanks!

ebrevdo commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -1046,13 +1054,19 @@ class biases.
+    if labels_as_indices:
+      # We want each row of labels to be the indices of the targets, which
+      # start at col 0 and end at col num_true-1.
+      out_labels = gen_array_ops.tile(
```
i guess this should only be used for num_true == 1, right?  can you add a check and corresponding documentation?

alternatively you could just move this code to sparse_sampled_softmax_loss and revert this.

TTrapper replied:
@drpngx just checking in because this has been sitting for some time. I responded to @ebrevdo 's review with the above comment. Thanks!

asimshankar commented:
(Removing API review label as this is in `contrib`)

TTrapper reviewed (commented):
@ebrevdo just checking in. I'm not sure whether my above comment triggers the necessary notifications to get the ball rolling again.

TTrapper committed:
Merge branch 'master' into sampled_sparse_softmax_loss

drpngx commented:
Thanks @TTrapper ! Let's see what @ebrevdo has to say.

ebrevdo commented:
If the special case code is only used by the sampled function, perhaps it's better to keep it in that function only?

TTrapper commented:
Thanks @ebrevdo. It's general enough to be potentially useful to others. I'm also weary of having code elsewhere that assumes `out_logits == concat[true_logits, sampled_logits]`. If that ever changes, then it is nice if changes to the `out_labels` are all in the same place as well.  Let me know if you agree.

TTrapper commented:
Hi @drpngx, @ebrevdo. Just checking in on my above comment. Thanks!

drpngx commented:
@TTrapper I think the recommendation was to move to contrib. That will still be available to everyone.

TTrapper committed:
Removing labels_as_indices logic from _compute_sampled_logits. Now computing 0-index labels in sampled_sparse_softmax_loss.

TTrapper committed:
calling array_ops instead of erroneus tf

TTrapper commented:
@drpngx Thanks, I think @ebrevdo 's concern was that the proposed `sample_sparse_softmax_loss` was the only place where the special case logic for getting true indices was used. I have now removed that special case logic from `_compute_sampled_logits` and added the specific use case directly to `sampled_sparse_softmax_loss`.

drpngx commented:
I see.

Jenkins, test this please.
